hbase-server/src/main/java/org/apache/hadoop/hbase/tool/BulkLoadHFilesTool.java

```diff
@@ -879,13 +881,21 @@ public void bulkHFile(ColumnFamilyDescriptorBuilder builder, FileStatus hfileSta
       }
 
       int maxRetries = getConf().getInt(HConstants.BULKLOAD_MAX_RETRIES_NUMBER, 10);
-      maxRetries = Math.max(maxRetries, startEndKeys.size() + 1);
+
+      /**
+       * For the first attempt, we make maxRetries with the configured maximum retry number
+       * As long as we find that region number changed, we setup maxRetries to region number
+       * But if we find that the region is not changed, then the maxRetries should be still
+       * be configured BULKLOAD_MAX_RETRIES_NUMBER to avoid meaningless retry attempts
+       */
+      if(count != 0 && previousRegionNum != startEndKeys.size() )
```

Comment:
This is a nice find.
Well this will help your particular case. But don't think this is a generic solution.
What if a cluster having an issue like your's (config issue) and during this configured retry times there was a split also!  It will change the max retries to be so large (if so many regions like ur case).
What if we just reset the 'count' only within the while loop if we notice a split happened in between the current run and previous? If there are splits happening at regular interval this will cause a never ending loop and would still need an upper bound. But blindly making the retry count to be #regions+1 just after seeing a split is something concerning to me.
cc @saintstack   - Seems you  reviewed the original jira so might be remembering that context. Any pointers sir?

Reply:
@anoopsjohn 
Yes, i agree with you , my above solution is not generic;
After totally review the whole calling path from my business code , to the HBase client retry logic and then finally to the server processing logic, I think things are clear now and the fix becomes more easier;
Firstly, in the RegionServer side, the critical unrecoverable exception is swallowed and I think this is incorrect:
[SecureBulkLoadManager.java#297](https://github.com/apache/hbase/blob/a9205f8f4d98ee672c0c6aa9cafa5ef2afc6aab5/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/SecureBulkLoadManager.java#L297), it eat all exceptions thrown from `region.bulkLoadHFiles`, this is obviously incorrect. I checked the commit history of this `catch` clause and it seems that it is a long-history commit. Whether or not exceptions are thrown will impact the Client side retry logic, which means, only when exception is thrown, the client side could get the error information(Refer [BulkLoadHFilesTool.java#396](https://github.com/apache/hbase/blob/a9205f8f4d98ee672c0c6aa9cafa5ef2afc6aab5/hbase-server/src/main/java/org/apache/hadoop/hbase/tool/BulkLoadHFilesTool.java#L397)
) and set up the retry policy;

Secondly, the exception in my incident is not processed correctly: [SecureBulkLoadManager.java#L395](https://github.com/apache/hbase/blob/a9205f8f4d98ee672c0c6aa9cafa5ef2afc6aab5/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/SecureBulkLoadManager.java#L395)
The exception information is 
`java.lang.IllegalArgumentException: Wrong FS: hdfs://warehousestore/tmp/mdm_user_segments_2798fd21-0961-43ff-8bd2-dcf0180c6918/h/d7a380a0cf1f4adbb459e9f0d0cf66c`
So I think all HDFS related code in the method should all be catched and thrown as IOException, instead of just rename() failed issue;

After above 2 issues are processed correctly, then client side will get the error information correctly at  [BulkLoadHFilesTool.java#396](https://github.com/apache/hbase/blob/a9205f8f4d98ee672c0c6aa9cafa5ef2afc6aab5/hbase-server/src/main/java/org/apache/hadoop/hbase/tool/BulkLoadHFilesTool.java#L397) and then  from the code, we could find that user could control the retry logic by configuration `hbase.client.retries.number` and `hbase.bulkload.retries.retryOnIOException`, instead of falling into retry disaster;